### ISSUE TITLE
Bug: Logic and arithmetic left shifts sometimes overflow the output signal

### ISSUE THREAD

**Issue by flaviens** [YosysHQ/yosys#4113]
### Version

Yosys 0.29+11 (git sha1 0b05a6425, gcc 7.5.0-6ubuntu2 -fPIC -Os), also tested on
Yosys 0.29+11 (git sha1 f7ffd12bf, gcc 7.5.0-6ubuntu2 -fPIC -Os)

### On which OS did this happen?

Linux

### Reproduction Steps

Hi there!

I encountered an issue for both logic and arithmetic left shifts. In some cases, it looks that the output may overflow.
For example, you can take this design:
```
module top(input logic [31:0] a, input logic [31:0] b, output logic [4:0] out_data);
  assign out_data = b[5:1] << a[31:27];
endmodule
```

And a testbench like:
```
int main() {
  uint32_t a = 0x4e57633b;
  uint32_t b = 0x468e7520;
  top.p_a.set<uint32_t>(a);
  top.p_b.set<uint32_t>(b);
  top.step();
  uint32_t out_data = top.p_out__data.get<uint32_t>();
  std::cout << "Out data: " << std::hex << out_data << std::endl;
}
```

Thank you!
Flavien

### Expected Behavior

Out data: `0`

### Actual Behavior

Out data: `2000` (hex)

**Comment by whitequark** [comment]
This behavior was intentional, but I agree that it is a bug.

**Comment by whitequark** [comment]
> Yosys 0.29+11 

Do note that you are six Yosys releases behind.

**Comment by povik** [comment]
> This behavior was intentional, but I agree that it is a bug.

Please elaborate. What aspect to the behavior was intentional?

**Comment by flaviens** [comment]
> > Yosys 0.29+11
> 
> Do note that you are six Yosys releases behind.

Thanks for the tip! I re-tested and updated the post accordingly.

**Comment by whitequark** [comment]
> What aspect to the behavior was intentional?

CXXRTL itself makes sure to mask the high bits of values at the point of use, rather than the point of computation, to avoid excessive and repeated masking. But `.get()` was never taught about that aspect. (This gets even worse when using the C API, so while `.get()` can and should be taught to clear the high bits, it may be a good idea to mask at the point of computation, so that consumers of the C API aren't burdened with this.)

**Comment by povik** [comment]
> CXXRTL itself makes sure to mask the high bits of values at the point of use, rather than the point of computation, to avoid excessive and repeated masking.

Huh, I thought values are supposed to be masked at the point of computation already, with that in mind I wrote e.g. #4075. So we are in the middle of a transition?

**Comment by whitequark** [comment]
> So we are in the middle of a transition?

Functionally yes. There is no harm in masking twice, so I've decided a while ago to blanket approve any changes that add masing at the point of computation. I didn't communicate it mainly due to the lack of time.

**Comment by whitequark** [comment]
@povik Do you have any appetite for adding the masking to `.get` and `.set` at least?

**Comment by whitequark** [comment]
@flaviens Not able to reproduce on commit fc5ff7a265293099e25812afbbbe8e687f0b62e8 on Linux x86_64 with clang++ or g++ `-O0` or `-O3`.

**Comment by whitequark** [comment]
This was fixed in commit ff53f3d2b6e267358cf4c3d08574416c79441ef9, I think.
